File: ios/StillMoment/Presentation/Views/GuidedMeditations/ContentGuideSheet.swift
//
//  ContentGuideSheet.swift
//  Still Moment
//
//  Presentation Layer - Curated, locale-specific list of free meditation sources.
//

import SwiftUI

/// Sheet listing curated, free meditation sources for the current locale.
///
/// Reachable from the empty-state secondary CTA and from the `info.circle`
/// button in the library nav bar. Source content lives in
/// `meditation_sources.json`; taps open the URL in the system browser.
struct ContentGuideSheet: View {
    // MARK: Lifecycle

    init(
        sources: [MeditationSource],
        onOpenURL: @escaping (URL) -> Void = { url in
            UIApplication.shared.open(url)
        },
        onDismiss: @escaping () -> Void
    ) {
        self.sources = sources
        self.onOpenURL = onOpenURL
        self.onDismiss = onDismiss
    }

    // MARK: Internal

    var body: some View {
        ZStack {
            self.theme.backgroundGradient
                .ignoresSafeArea()

            ScrollView {
                VStack(alignment: .leading, spacing: 0) {
                    self.titleRow
                    self.intro
                    self.importBanners
                    self.sourceList
                }
                .padding(.horizontal, 22)
                .padding(.bottom, 24)
            }
        }
        .accessibilityIdentifier("library.guideSheet")
    }

    // MARK: Private

    @Environment(\.themeColors)
    private var theme

    private let sources: [MeditationSource]
    private let onOpenURL: (URL) -> Void
    private let onDismiss: () -> Void

    private var titleRow: some View {
        HStack(alignment: .center) {
            Text("guided_meditations.guide.title")
                .themeFont(.screenTitle)
                .accessibilityAddTraits(.isHeader)
            Spacer()
            Button(action: self.onDismiss) {
                Image(systemName: "xmark")
                    .font(.system(size: 14, weight: .semibold))
                    .foregroundColor(self.theme.textSecondary)
                    .frame(width: 30, height: 30)
                    .background(
                        Circle().fill(self.theme.cardBackground.opacity(.opacitySecondary))
                    )
            }
            .accessibilityLabel("guided_meditations.guide.close")
            .accessibilityIdentifier("library.guideSheet.close")
        }
        .padding(.top, 8)
        .padding(.bottom, 10)
    }

    private var intro: some View {
        Text("guided_meditations.guide.intro")
            .themeFont(.bodySecondary, color: \.textSecondary)
            .padding(.bottom, 24)
    }

    private var importBanners: some View {
        VStack(spacing: 10) {
            NavigationLink {
                HowToImportBrowserView()
            } label: {
                ImportBannerCard(
                    icon: "safari",
                    titleKey: "guided_meditations.guide.banner.browser.title",
                    subtitleKey: "guided_meditations.guide.banner.browser.subtitle"
                )
            }
            .buttonStyle(.plain)
            .accessibilityIdentifier("library.guideSheet.banner.browser")

            NavigationLink {
                HowToImportFilesView()
            } label: {
                ImportBannerCard(
                    icon: "folder",
                    titleKey: "guided_meditations.guide.banner.files.title",
                    subtitleKey: "guided_meditations.guide.banner.files.subtitle"
                )
            }
            .buttonStyle(.plain)
            .accessibilityIdentifier("library.guideSheet.banner.files")
        }
        .padding(.bottom, 24)
    }

    private var sourceList: some View {
        VStack(spacing: 0) {
            ForEach(Array(self.sources.enumerated()), id: \.element.id) { index, source in
                SourceRow(
                    source: source,
                    showsTopDivider: index > 0
                ) {
                    self.handleTap(on: source)
                }
            }
        }
        .background(
            RoundedRectangle(cornerRadius: 24)
                .fill(self.theme.cardBackground.opacity(.opacitySecondary))
        )
        .overlay(
            RoundedRectangle(cornerRadius: 24)
                .strokeBorder(self.theme.cardBorder, lineWidth: 0.5)
        )
    }

    private func handleTap(on source: MeditationSource) {
        self.onOpenURL(source.url)
        self.onDismiss()
    }
}

// MARK: - Banner Card

private struct ImportBannerCard: View {
    // MARK: Internal

    let icon: String
    let titleKey: String
    let subtitleKey: String

    var body: some View {
        HStack(alignment: .center, spacing: 14) {
            self.iconBubble
            self.text
            Spacer(minLength: 8)
            Image(systemName: "chevron.right")
                .font(.system(size: 14, weight: .semibold))
                .foregroundColor(self.theme.textSecondary)
                .opacity(.opacitySecondary)
        }
        .padding(14)
        .frame(maxWidth: .infinity, alignment: .leading)
        .background(
            RoundedRectangle(cornerRadius: 18)
                .fill(self.theme.accentBannerBackground)
        )
        .overlay(
            RoundedRectangle(cornerRadius: 18)
                .strokeBorder(self.theme.accentBannerBorder, lineWidth: 1)
        )
        .contentShape(Rectangle())
        .accessibilityElement(children: .ignore)
        .accessibilityLabel(self.accessibilityLabel)
        .accessibilityAddTraits(.isButton)
    }

    // MARK: Private

    @Environment(\.themeColors)
    private var theme

    private var iconBubble: some View {
        ZStack {
            Circle()
                .fill(self.theme.accentBubbleBackground)
                .frame(width: 36, height: 36)
            Image(systemName: self.icon)
                .font(.system(size: 18, weight: .regular))
                .foregroundColor(self.theme.interactive)
        }
    }

    private var text: some View {
        VStack(alignment: .leading, spacing: 3) {
            Text(LocalizedStringKey(self.titleKey))
                .themeFont(.listTitle)
                .multilineTextAlignment(.leading)
            Text(LocalizedStringKey(self.subtitleKey))
                .themeFont(.bodySecondary, color: \.textSecondary)
                .multilineTextAlignment(.leading)
                .fixedSize(horizontal: false, vertical: true)
        }
    }

    private var accessibilityLabel: String {
        let title = NSLocalizedString(self.titleKey, comment: "")
        let subtitle = NSLocalizedString(self.subtitleKey, comment: "")
        return "\(title), \(subtitle)"
    }
}

// MARK: - Row

private struct SourceRow: View {
    // MARK: Internal

    let source: MeditationSource
    let showsTopDivider: Bool
    let onTap: () -> Void

    var body: some View {
        Button(action: self.onTap) {
            HStack(alignment: .top, spacing: 12) {
                VStack(alignment: .leading, spacing: 4) {
                    self.titleLine
                    Text(self.source.description)
                        .themeFont(.bodySecondary, color: \.textSecondary)
                        .multilineTextAlignment(.leading)
                    Text(self.source.host)
                        .themeFont(.caption, color: \.textSecondary)
                        .opacity(.opacitySecondary)
                }
                Spacer(minLength: 12)
                Image(systemName: "arrow.up.right.square")
                    .font(.system(size: 18, weight: .regular))
                    .foregroundColor(self.theme.interactive)
                    .padding(.top, 2)
            }
            .padding(.horizontal, 16)
            .padding(.vertical, 14)
            .frame(maxWidth: .infinity, alignment: .leading)
            .contentShape(Rectangle())
        }
        .buttonStyle(.plain)
        .overlay(alignment: .top) {
            if self.showsTopDivider {
                Rectangle()
                    .fill(self.theme.cardBorder.opacity(.opacitySecondary))
                    .frame(height: 0.5)
                    .padding(.horizontal, 12)
            }
        }
        .accessibilityElement(children: .combine)
        .accessibilityLabel(self.accessibilityLabel)
        .accessibilityHint("guided_meditations.guide.openSource")
        .accessibilityAddTraits(.isLink)
        .accessibilityIdentifier("library.guideSheet.row.\(self.source.id)")
    }

    // MARK: Private

    @Environment(\.themeColors)
    private var theme

    private var accessibilityLabel: String {
        var parts = [self.source.name]
        if let author = source.author {
            parts.append(author)
        }
        parts.append(self.source.description)
        return parts.joined(separator: ", ")
    }

    @ViewBuilder private var titleLine: some View {
        if let author = self.source.author {
            HStack(alignment: .firstTextBaseline, spacing: 6) {
                Text(self.source.name)
                    .themeFont(.listTitle)
                Text("·")
                    .themeFont(.bodySecondary, color: \.textSecondary)
                Text(author)
                    .themeFont(.bodySecondary, color: \.textSecondary)
            }
        } else {
            Text(self.source.name)
                .themeFont(.listTitle)
        }
    }
}

// MARK: - Previews

#if DEBUG
private let previewSources: [MeditationSource] = [
    MeditationSource(
        id: "tara-brach",
        name: "Tara Brach",
        author: nil,
        description: "Guided meditations, RAIN practice. Direct MP3.",
        host: "tarabrach.com",
        // swiftlint:disable:next force_unwrapping
        url: URL(string: "https://www.tarabrach.com/guided-meditations/")!
    ),
    MeditationSource(
        id: "audio-dharma",
        name: "Audio Dharma",
        author: "Gil Fronsdal",
        description: "Vipassana tradition. Direct MP3.",
        host: "audiodharma.org",
        // swiftlint:disable:next force_unwrapping
        url: URL(string: "https://www.audiodharma.org/")!
    )
]

@available(iOS 17.0, *)
#Preview("Guide Sheet") {
    Color.clear
        .sheet(isPresented: .constant(true)) {
            ThemeRootView {
                NavigationStack {
                    ContentGuideSheet(
                        sources: previewSources,
                        onOpenURL: { _ in },
                        onDismiss: {}
                    )
                }
            }
            .presentationDetents([.large])
            .presentationDragIndicator(.visible)
        }
}
#endif

File: android/app/src/main/kotlin/com/stillmoment/presentation/ui/meditations/ContentGuideSheet.kt
package com.stillmoment.presentation.ui.meditations

import android.content.Intent
import android.net.Uri
import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.verticalScroll
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.automirrored.filled.OpenInNew
import androidx.compose.material3.ExperimentalMaterial3Api
import androidx.compose.material3.Icon
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.ModalBottomSheet
import androidx.compose.material3.Text
import androidx.compose.material3.rememberModalBottomSheetState
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.platform.LocalConfiguration
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.semantics.Role
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.heading
import androidx.compose.ui.semantics.role
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.tooling.preview.Preview
import androidx.compose.ui.unit.dp
import com.stillmoment.R
import com.stillmoment.domain.models.MeditationSource
import com.stillmoment.presentation.ui.theme.StillMomentTheme
import com.stillmoment.presentation.ui.theme.TypographyRole
import com.stillmoment.presentation.ui.theme.textColor
import com.stillmoment.presentation.ui.theme.textStyle
import kotlinx.collections.immutable.ImmutableList
import kotlinx.collections.immutable.persistentListOf

/**
 * Modal bottom sheet listing curated, free meditation sources for the current locale.
 *
 * Reachable from the empty-state secondary CTA and from the info icon in the
 * library top app bar. Source content lives in `assets/meditation_sources.json`;
 * taps open the URL in the system browser.
 */
@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun ContentGuideSheet(
    sources: ImmutableList<MeditationSource>,
    onDismiss: () -> Unit,
    modifier: Modifier = Modifier,
    onOpenUrl: ((String) -> Unit)? = null
) {
    val context = LocalContext.current
    val openHandler: (String) -> Unit = onOpenUrl ?: { url ->
        val intent = Intent(Intent.ACTION_VIEW, Uri.parse(url))
        intent.flags = Intent.FLAG_ACTIVITY_NEW_TASK
        runCatching { context.startActivity(intent) }
    }

    ModalBottomSheet(
        onDismissRequest = onDismiss,
        sheetState = rememberModalBottomSheetState(skipPartiallyExpanded = true),
        containerColor = MaterialTheme.colorScheme.surface,
        modifier = modifier
    ) {
        ContentGuideSheetContent(
            sources = sources,
            onSourceClick = { source ->
                openHandler(source.url)
                onDismiss()
            }
        )
    }
}

@Composable
internal fun ContentGuideSheetContent(
    sources: ImmutableList<MeditationSource>,
    onSourceClick: (MeditationSource) -> Unit,
    modifier: Modifier = Modifier
) {
    Column(
        modifier = modifier
            .fillMaxWidth()
            .verticalScroll(rememberScrollState())
            .padding(horizontal = 22.dp)
            .padding(bottom = 32.dp)
    ) {
        Text(
            text = stringResource(R.string.guided_meditations_guide_title),
            style = TypographyRole.ScreenTitle.textStyle(),
            color = TypographyRole.ScreenTitle.textColor(),
            modifier = Modifier.semantics { heading() }
        )

        Spacer(modifier = Modifier.height(10.dp))

        Text(
            text = stringResource(R.string.guided_meditations_guide_intro),
            style = TypographyRole.BodySecondary.textStyle(),
            color = TypographyRole.BodySecondary.textColor()
        )

        Spacer(modifier = Modifier.height(24.dp))

        SourceCard(sources = sources, onSourceClick = onSourceClick)
    }
}

@Composable
private fun SourceCard(sources: ImmutableList<MeditationSource>, onSourceClick: (MeditationSource) -> Unit) {
    val borderColor = MaterialTheme.colorScheme.onSurface.copy(alpha = 0.08f)
    Column(
        modifier = Modifier
            .fillMaxWidth()
            .clip(RoundedCornerShape(24.dp))
            .background(MaterialTheme.colorScheme.surfaceVariant.copy(alpha = 0.4f))
            .border(0.5.dp, borderColor, RoundedCornerShape(24.dp))
    ) {
        sources.forEachIndexed { index, source ->
            if (index > 0) {
                Box(
                    modifier = Modifier
                        .fillMaxWidth()
                        .padding(horizontal = 12.dp)
                        .height(0.5.dp)
                        .background(borderColor)
                )
            }
            SourceRow(source = source, onClick = { onSourceClick(source) })
        }
    }
}

@Composable
private fun SourceRow(source: MeditationSource, onClick: () -> Unit) {
    val rowDescription = buildString {
        append(source.name)
        source.author?.let {
            append(", ")
            append(it)
        }
        append(", ")
        append(source.description)
    }
    val openLabel = stringResource(R.string.guided_meditations_guide_open_source)

    Row(
        modifier = Modifier
            .fillMaxWidth()
            .clickable(role = Role.Button, onClick = onClick)
            .semantics {
                role = Role.Button
                contentDescription = "$rowDescription. $openLabel"
            }
            .padding(horizontal = 16.dp, vertical = 14.dp),
        verticalAlignment = Alignment.Top
    ) {
        Column(modifier = Modifier.weight(1f)) {
            SourceTitleLine(source = source)
            Spacer(modifier = Modifier.height(4.dp))
            Text(
                text = source.description,
                style = TypographyRole.BodySecondary.textStyle(),
                color = TypographyRole.BodySecondary.textColor()
            )
            Spacer(modifier = Modifier.height(4.dp))
            Text(
                text = source.host,
                style = TypographyRole.Caption.textStyle(),
                color = TypographyRole.Caption.textColor().copy(alpha = 0.7f)
            )
        }
        Spacer(modifier = Modifier.width(12.dp))
        Icon(
            imageVector = Icons.AutoMirrored.Filled.OpenInNew,
            contentDescription = null,
            tint = MaterialTheme.colorScheme.primary,
            modifier = Modifier.size(20.dp)
        )
    }
}

@Composable
private fun SourceTitleLine(source: MeditationSource) {
    val author = source.author
    if (author == null) {
        Text(
            text = source.name,
            style = TypographyRole.ListTitle.textStyle(),
            color = TypographyRole.ListTitle.textColor()
        )
    } else {
        Row(verticalAlignment = Alignment.Bottom) {
            Text(
                text = source.name,
                style = TypographyRole.ListTitle.textStyle(),
                color = TypographyRole.ListTitle.textColor()
            )
            Spacer(modifier = Modifier.width(6.dp))
            Text(
                text = "·",
                style = TypographyRole.BodySecondary.textStyle(),
                color = TypographyRole.BodySecondary.textColor()
            )
            Spacer(modifier = Modifier.width(6.dp))
            Text(
                text = author,
                style = TypographyRole.BodySecondary.textStyle(),
                color = TypographyRole.BodySecondary.textColor()
            )
        }
    }
}

// MARK: - Locale helper

/** Returns the language code (`"de"`, `"en"`, ...) for the active app configuration. */
@Composable
fun currentLanguageCode(): String = LocalConfiguration.current.locales[0].language.ifBlank { "en" }

// MARK: - Previews

@Preview(showBackground = true, name = "Guide Sheet (DE)")
@Composable
private fun ContentGuideSheetPreview() {
    val sources = persistentListOf(
        MeditationSource(
            id = "tara-brach",
            name = "Tara Brach",
            author = null,
            description = "Guided meditations, RAIN practice. Direct MP3.",
            host = "tarabrach.com",
            url = "https://www.tarabrach.com/guided-meditations/"
        ),
        MeditationSource(
            id = "audio-dharma",
            name = "Audio Dharma",
            author = "Gil Fronsdal",
            description = "Vipassana tradition. Direct MP3.",
            host = "audiodharma.org",
            url = "https://www.audiodharma.org/"
        )
    )
    StillMomentTheme {
        Box(modifier = Modifier.background(Color.Black)) {
            ContentGuideSheetContent(sources = sources, onSourceClick = {})
        }
    }
}
